kernel/src/table_changes/log_replay.rs

```diff
@@ -0,0 +1,624 @@
+            //    actions.as_ref().any_ref().downcast_ref().unwrap();
+            //println!("actions: {:?}", actions_arrow.record_batch());
+            // apply data skipping to get back a selection vector for actions that passed skipping
+            // note: None implies all files passed data skipping.
+            let filter_vector = self
+                .filter
+                .as_ref()
+                .map(|filter| filter.apply(actions.as_ref()))
+                .transpose()?;
+
+            // we start our selection vector based on what was filtered. we will add to this vector
+            // below if a file has been removed
+            let selection_vector = match filter_vector {
+                Some(ref filter_vector) => filter_vector.clone(),
+                None => vec![true; actions.len()],
+            };
+            let mut visitor = Phase1Visitor::new(self, selection_vector);
+            visitor.visit_rows_of(actions.as_ref())?;
+        }
+        Ok(())
+    }
+
+    fn resolve_dvs(&mut self) {
+        if self.has_cdc_action {
+            self.remove_dvs.clear();
+        } else {
+            self.remove_dvs
+                .retain(|rm_path, _| self.add_paths.contains(rm_path));
+        }
+        self.add_paths.clear()
+    }
+
+    fn into_scan_batches(
+        self,
+    ) -> DeltaResult<impl Iterator<Item = DeltaResult<TableChangesScanData>>> {
+        let Self {
+            has_cdc_action,
+            remove_dvs,
+            commit_file,
+            json_client,
+            timestamp: _,
+            filter,
+            expression_evaluator,
+            add_paths: _,
+        } = self;
+        let remove_dvs = Arc::new(remove_dvs);
+
+        let schema = Phase2Visitor::schema()?;
+        let action_iter =
+            json_client.read_json_files(&[commit_file.location.clone()], schema, None)?;
+
+        let result = action_iter.map(move |actions| -> DeltaResult<_> {
+            let actions = actions?;
+            // apply data skipping to get back a selection vector for actions that passed skipping
+            // note: None implies all files passed data skipping.
+            let filter_vector = filter
+                .as_ref()
+                .map(|filter| filter.apply(actions.as_ref()))
+                .transpose()?;
+
+            // we start our selection vector based on what was filtered. we will add to this vector
+            // below if a file has been removed
+            let selection_vector = match filter_vector {
+                Some(ref filter_vector) => filter_vector.clone(),
+                None => vec![true; actions.len()],
+            };
+            let mut visitor = Phase2Visitor::new(&remove_dvs, selection_vector, has_cdc_action);
+            visitor.visit_rows_of(actions.as_ref())?;
+
+            let result = expression_evaluator.evaluate(actions.as_ref())?;
+            Ok((result, visitor.selection_vector, remove_dvs.clone()))
```

Comment:
Can we make `remove_dvs` an `Arc<HashMap>` so we don't have to clone the whole `HashMap`